### PR TITLE
Playback Scrubber

### PR DESCRIPTION
**Pull request by JShep1**
This is a base implementation draft mainly to get feedback on the gui portion.  Below is a small demo with only the scrubber bar itself.  Any enhancements or usability improvement recommendations would be appreciated. @coleosrf @nkoenig @chapulina 
playbackscrubber

So far I am going to add:

- Start and end times on either side of the slide bar
- The current time displayed above the slider marker
- Manually enter time to jump to

### CODE REVIEW

**Comment by ColeOSRF** [comment]
After you add the current time it might be good to be able to click on that and type in a specific time. Could be useful for getting the time right where you want it without having to fiddle with the slider to get small increments. Speaking of small increments maybe an arrow on each side of the slider that you can click on and it will move the time by a small increment. Not sure the best way to determine how big it should jump each time you press the arrow. Perhaps a percentage of the time? Maybe the user should be able to set the increments?

**Comment by JShep1** [comment]
> After you add the current time it might be good to be able to click on that and type in a specific time. Could be useful for getting the time right where you want it without having to fiddle with the slider to get small increments. Speaking of small increments maybe an arrow on each side of the slider that you can click on and it will move the time by a small increment. Not sure the best way to determine how big it should jump each time you press the arrow. Perhaps a percentage of the time? Maybe the user should be able to set the increments?

Yeah entering the time would definitely be useful.  I think we're going to rely on the World Control buttons on the lower left to control the stepping through the log playback.  Perhaps this prompts a step back button to accompany the play and step forward buttons

**Comment by JShep1** [comment]
logplayback
Ready for another review, here's a preview

**Comment by JShep1** [comment]
Depends on https://github.com/ignitionrobotics/ign-math/pull/152

File: src/cmd/cmdgazebo.rb.in
Comment (chapulina) [inline src/cmd/cmdgazebo.rb.in:367]: I think it would be good to print a warning to the user in case they specified `--gui-config`.

Or even better, don't change the `gui_config` if the user manually specified it. They may be wanting to playback with a specific config.

File: src/gui/Gui.cc
Comment (chapulina) [inline src/gui/Gui.cc:164]: I think it should be possible to reduce some duplication moving some logic to the block below?

File: src/gui/plugins/playback_scrubber/PlaybackScrubber.cc
Comment (chapulina) [inline src/gui/plugins/playback_scrubber/PlaybackScrubber.cc:196]: We shouldn't hardcode the world name here, you should be able to get it from [ignition::gui::worldNames()](https://ignitionrobotics.org/api/gui/3.3/namespaceignition_1_1gui.html#a16c452c451b1405a336472b2ed6a9c01).

File: src/gui/plugins/playback_scrubber/PlaybackScrubber.cc
Comment (chapulina) [inline src/gui/plugins/playback_scrubber/PlaybackScrubber.cc:212]: Same thing about hardcoding the world name here

File: src/gui/plugins/playback_scrubber/PlaybackScrubber.hh
Comment (chapulina) [inline src/gui/plugins/playback_scrubber/PlaybackScrubber.hh:77]: This threw me off for a second. Reading the description, I'd expect this to be `OnDrop` :upside_down_face:

File: src/gui/plugins/playback_scrubber/PlaybackScrubber.cc
Comment (JShep1) [reply src/gui/plugins/playback_scrubber/PlaybackScrubber.cc:196]: So I'm not sure if this was a bug or not.  If I get the proper world name and use it, the playback callback doesn't work, but if I use `default` for the world name it works.

File: src/cmd/cmdgazebo.rb.in
Comment (JShep1) [reply src/cmd/cmdgazebo.rb.in:368]: I just added a condition to not utilize the playback config if the `--gui-config` flag is specified here 924af51600d0083aeadfd8b682c0b9d39a157bb2

File: src/gui/Gui.cc
Comment (JShep1) [reply src/gui/Gui.cc]: You're right, not sure why I didn't do this before 924af51600d0083aeadfd8b682c0b9d39a157bb2

File: src/gui/plugins/playback_scrubber/PlaybackScrubber.hh
Comment (JShep1) [reply src/gui/plugins/playback_scrubber/PlaybackScrubber.hh]: Thanks for the reminder, I meant to change this from an earlier iteration 924af51600d0083aeadfd8b682c0b9d39a157bb2

File: src/gui/plugins/playback_scrubber/PlaybackScrubber.cc
Comment (JShep1) [reply src/gui/plugins/playback_scrubber/PlaybackScrubber.cc]: Resolved e7a428e1e3fbd4c02f4f17a14817dcadf989b751

File: src/gui/plugins/playback_scrubber/PlaybackScrubber.cc
Comment (JShep1) [reply src/gui/plugins/playback_scrubber/PlaybackScrubber.cc]: Resolved e7a428e1e3fbd4c02f4f17a14817dcadf989b751

File: src/gui/Gui.cc
Comment (chapulina) [inline src/gui/Gui.cc:161]: I think you need to move this logic up, before the App calls `SetDefaultConfigPath`

File: src/gui/plugins/playback_scrubber/PlaybackScrubber.qml
Comment (chapulina) [inline src/gui/plugins/playback_scrubber/PlaybackScrubber.qml:153]: I noticed that once I input something in the text field, it stops being updated when I play or seek the simulation. It starts updating again if I delete the text. Is that intentional?

File: src/systems/log/LogPlayback.cc
Comment (chapulina) [inline src/systems/log/LogPlayback.cc:305]: Could you add a check on `test/integration/log_system.cc` that verifies that the component was correctly created and has the correct values?

File: src/gui/plugins/playback_scrubber/PlaybackScrubber.qml
Comment (chapulina) [inline src/gui/plugins/playback_scrubber/PlaybackScrubber.qml:52]: Could you document the functions and the properties? See an example [here](https://github.com/ignitionrobotics/ign-gazebo/blob/master/src/gui/plugins/component_inspector/ComponentInspector.qml).

**Review by chapulina — COMMENTED** [review]
Works for me! All the different font sizes, colors and weights in the widget threw me off a bit, I think it would be good to be consistent.

I also feel like there's too much redundant information. The current time and end times are displayed twice. 

One suggestion:

* Remove the `Current time` line on the top
* Replace the start time with the current time input field (gotta make sure it updates even after edited)
* Remove the `/ end time`

We lose the start time, but the user can always scroll back to it if they want? We could also display it on a tool tip?

> we're going to rely on the World Control buttons on the lower left

How about placing the scrubber next to it so they look like they're the same widget, without necessarily needing to be? I think you could anchor them together.

>  Perhaps this prompts a step back button to accompany the play and step forward buttons

I think that's a good idea. If it's too much to add it now, I'd recommend ticketing an issue in case someone is motivated to contribute. At least for now, that button should only show up for log playback.

**Comment by nkoenig** [comment]
> * Remove the `Current time` line on the top

I second this.

> * Replace the start time with the current time input field (gotta make sure it updates even after edited)
> * Remove the `/ end time`

How about removing the start time and end time from both ends of the slider and leave  the `<current_time_input> / end_time` below the scrubber. I find the `end_time` useful in order to enter a correct time in the input field. 

See #354 for other improvement to the time input field. This can be saved for later.

**Comment by chapulina** [comment]
> How about removing the start time and end time from both ends of the slider and leave the <current_time_input> / end_time below the scrubber.

File: src/gui/plugins/playback_scrubber/PlaybackScrubber.qml
Comment (JShep1) [reply src/gui/plugins/playback_scrubber/PlaybackScrubber.qml:153]: Yes, I thought about updating it, but then thought a user might enter a particular time value and keep wanting to go back to it as opposed to having to enter it after every time they play the log back.  Do you think we should just remove the current time preview in the text field?

**Comment by JShep1** [comment]
New fixes are in, the plugin now looks like this
Screenshot from 2020-09-16 15-15-59

File: src/gui/plugins/playback_scrubber/PlaybackScrubber.qml
Comment (JShep1) [reply src/gui/plugins/playback_scrubber/PlaybackScrubber.qml:81]: 39ca2b20e1f33e08d173513b83b827e692b73f82

File: src/systems/log/LogPlayback.cc
Comment (JShep1) [reply src/systems/log/LogPlayback.cc:305]: 39ca2b20e1f33e08d173513b83b827e692b73f82

File: src/gui/Gui.cc
Comment (JShep1) [reply src/gui/Gui.cc]: 39ca2b20e1f33e08d173513b83b827e692b73f82

File: src/gui/plugins/playback_scrubber/PlaybackScrubber.qml
Comment (JShep1) [reply src/gui/plugins/playback_scrubber/PlaybackScrubber.qml]: With the new QML updates, your suggestion makes more sense given the empty text box is now the way of determining the current time so I updated it here a6d63aebcbab46eb7e4b99bdd82cd4857f9f6dc0

**Review by nkoenig — APPROVED** [review]

File: src/gui/Gui.cc
Comment (chapulina) [inline src/gui/Gui.cc:159]: I think this block can't be removed. Passing a custom `--gui-config` isn't working right now, either during simulation or playback.

**Review by chapulina — CHANGES_REQUESTED** [review]

File: src/gui/Gui.cc
Comment (JShep1) [reply src/gui/Gui.cc]: Thanks for catching this, I don't believe I meant to remove it, I added it back 13dd0e626c31a43529d2078830e8601d998094bd

**Comment by adlarkin** [comment]
I just wanted to comment that this seems to work well with #351. This is definitely a great addition, thanks @JShep1!

**Review by chapulina — APPROVED** [review]

**Comment by chapulina** [comment]
The `INTEGRATION_log_system` test failures look suspicious:

```
2020-09-17T19:34:09.1303931Z /github/workspace/test/integration/log_system.cc:745: Failure
2020-09-17T19:34:09.1304402Z Value of: common::exists(logFile)
2020-09-17T19:34:09.1304708Z   Actual: false
2020-09-17T19:34:09.1304964Z Expected: true
2020-09-17T19:34:09.1305924Z [1;33m[Wrn] [Filesystem.cc:288] [0m[1;33mFailed to open file [[0m[1;33m/github/workspace/build/test/test_logs/test_logs_record/state.tlog[0m[1;33m]: [0m[1;33mNo such file or directory[0m[1;33m[0m
2020-09-17T19:34:09.1306845Z /github/workspace/test/integration/log_system.cc:750: Failure
2020-09-17T19:34:09.1307359Z Value of: common::exists(logPlaybackFile)
2020-09-17T19:34:09.1307721Z   Actual: false
2020-09-17T19:34:09.1307963Z Expected: true
2020-09-17T19:34:09.1308316Z Failed to open the requested sqlite3 database
2020-09-17T19:34:09.1308809Z /github/workspace/test/integration/log_system.cc:773: Failure
2020-09-17T19:34:09.1309713Z Expected: (batch.end()) != (recordedIter), actual: 8-byte object <90-85 47-7B B7-55 00-00> vs 8-byte object <A0-2D 00-8C 5E-7F 00-00>
```

**Comment by JShep1** [comment]
> The `INTEGRATION_log_system` test failures look suspicious:
Agreed, is this a flaky test?  Jenkins shows it as not being a new failure, unless I'm interpreting it wrong.

**Comment by chapulina** [comment]
> Agreed, is this a flaky test? Jenkins shows it as not being a new failure

If you look here, it seems to be flaky because it's failed before:

https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/4177/testReport/junit/(root)/INTEGRATION_log_system/test_ran/history/

But if you look into each of those builds, they've all been builds related to this PR. 

If you look at the history of this test for the `ign-gazebo2` branch, it looks like it's been passing consistently:

https://build.osrfoundation.org/job/ignition_gazebo-ci-ign-gazebo2-bionic-amd64/149/testReport/(root)/LogSystemTest/history/

I haven't tried reproducing the failure locally yet, but I think we should take a look.

**Comment by chapulina** [comment]
I can reproduce the failures locally and the latest `ign-gazebo2` build that I just triggered doesn't have them. I'm looking into it.

https://build.osrfoundation.org/job/ignition_gazebo-ci-ign-gazebo2-bionic-amd64/150/

**Comment by chapulina** [comment]
The changes in af1a070 fixed the tests for me locally. I have no idea why that would be an issue on this PR but not on `ign-gazebo2` though :woman_shrugging:

**Comment by chapulina** [comment]
Phew, so it looks like `log_system` is fixed. And I just noticed that `Gui_TEST` is also a new failure that I can reproduce locally. On it.

**Comment by chapulina** [comment]
Ok, `Gui_TEST` is fixed in a7fbd0c.

**Comment by JShep1** [comment]
Thanks for fixing those issues!  I'm assuming we'll be good to merge once CI runs now?

**Comment by chapulina** [comment]
Argh, it looks like `log_system` has been passing consistently on Mac:

https://build.osrfoundation.org/job/ignition_gazebo-ci-ign-gazebo2-homebrew-amd64/156/testReport/(root)/LogSystemTest/history/

And I see it passed on previous commits of this PR. Let's run it one more time just to sanity check. I retriggered it.

**Comment by chapulina** [comment]
Mac passed now! Merging! :tada:

**Pull request merged**
